### DISCUSSION TITLE
Eglot Panics When Combined with Electric Pair Local mode

### DISCUSSION THREAD

**Discussion by creese** [joaotavora/eglot#1315] [Q&A]
I've been experimenting with Eglot as my primary tool and ran into an unexpected error. I'd appreciate any insights or potential solutions.

Here's a step-by-step of what I did:

1. Install rust-analyzer.
2. Add a hook to ensure Eglot is enabled for Rust mode.
3. Create a new project using `cargo new hello_cargo`.
4. Below the default `println!`, add another `println!`.
5. Upon typing the opening parenthesis for the second `println!()`, there is no error.
6. Enable `electric-pair-local-mode`.
7. Add another `println!`.
8. Upon typing the opening parenthesis for the third `println!()`, the following error message appears:

`*Messages*`:
```
jsonrpc-error: "request id=7 failed:", (jsonrpc-error-code . -32603), (jsonrpc-error-message . "request handler panicked: assertion failed: !!text[usize::from(position.offset)..].starts_with(char_typed)"), (jsonrpc-error-data)
```

`eglot-stderr-buffer`: 
```
Panic context:
> 
version: 0.0.0 (6572ec8d9 2023-10-15)
request: textDocument/onTypeFormatting DocumentOnTypeFormattingParams {
    text_document_position: TextDocumentPositionParams {
        text_document: TextDocumentIdentifier {
            uri: Url {
                scheme: "file",
                cannot_be_a_base: false,
                username: "",
                password: None,
                host: None,
                port: None,
                path: "/Source/hello_cargo/src/main.rs",
                query: None,
                fragment: None,
            },
        },
        position: Position {
            line: 3,
            character: 14,
        },
    },
    ch: "(",
    options: FormattingOptions {
        tab_size: 8,
        insert_spaces: true,
        properties: {},
        trim_trailing_whitespace: None,
        insert_final_newline: Some(
            true,
        ),
        trim_final_newlines: Some(
            true,
        ),
    },
}

thread 'LspServer' panicked at crates/rust-analyzer/src/handlers/request.rs:339:8:
assertion failed: !!text[usize::from(position.offset)..].starts_with(char_typed)
stack backtrace:
   0: rust_begin_unwind
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:595:5
   1: core::panicking::panic_fmt
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:67:14
   2: rust_analyzer::handlers::request::handle_on_type_formatting
   3: std::panicking::try
   4: rust_analyzer::dispatch::RequestDispatcher::on_sync
   5: rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::on_request
   6: rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::run
   7: rust_analyzer::main_loop::main_loop
   8: rust_analyzer::run_server
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Has anyone encountered a similar issue or have suggestions for a fix?

For reference, here's my configuration:
```
(add-hook 'rust-ts-mode-hook 'eglot-ensure)
(add-hook 'prog-mode-hook 'electric-pair-local-mode)
```

Thanks!

**Comment by joaotavora** [comment]
Converting to discussion. Not a MRE.

**Comment by joaotavora** [comment]
Seems very clearly like a server panic, not a client panic to me.

If still you want me to try to reproduce it, follow the steps in https://joaotavora.github.io/eglot/#Reporting-bugs

**Comment by creese** [comment]
Clearly, there's an issue with the server. However, when I switch to LSP Mode, the issue doesn't arise. It could be related to how the client and server are communicating. I'll take a look at the steps you mentioned.

**Comment by joaotavora** [comment]
> However, when I switch to LSP Mode, the issue doesn't arise.

If I had to guess, I would guess that LSP mode doesn't exercise the `textDocument/onTypeFormatting` capability of the server in Rust buffers, so the bug in the server doesn't have a chance to manifest itself.

I suspect the language server to be confused because soon after telling it that a `(`  has been inserted, which would presumably lead the server to also ask the client to insert a `)` , Emacs itself inserts the `)`.  This somehow confuses the server.

As a workaround, if you want Eglot to stop using this capability in `rust-mode` buffers:

```
(add-hook 'rust-mode-hook 
  (lambda () (setq eglot-ignored-server-capabilities 
    (delete :documentOnTypeFormattingProvider eglot-ignored-server-capabilities)))
```

This really should be reported to the rust-analyzer folks, though. 

I've tested this in a simple hello world program with rust-analyzer (`rust-analyzer 1 (6572ec8d94 2023-10-15)`) and couldn't reproduce. 

I recommend you try to make a recipe starting from `Emacs -Q` with nothing but `rust-mode`, `rust-analyzer` and `electric-pair-local-mode`.  NO `eglot-ensure` hookage needed.
